common/src/main/java/com/revenuecat/purchases/common/networking/ETagManager.kt

```diff
@@ -0,0 +1,120 @@
+package com.revenuecat.purchases.common.networking
+
+import android.content.Context
+import android.content.SharedPreferences
+import com.revenuecat.purchases.common.sha1
+import org.json.JSONObject
+import java.net.HttpURLConnection
+
+private const val NOT_MODIFIED_RESPONSE_CODE = 304
+private const val INTERNAL_SERVER_ERROR_RESPONSE_CODE = 500
+private typealias RequestHash = String
+
+private const val SERIALIZATION_NAME_ETAG = "eTag"
+private const val SERIALIZATION_NAME_HTTPRESULT = "httpResult"
+
+internal const val ETAG_HEADER_NAME = "X-RevenueCat-ETag"
+
+data class HTTPResultWithETag(
+    val eTag: String,
+    val httpResult: HTTPResult
+) {
+    fun serialize(): String {
+        return JSONObject().apply {
+            put(SERIALIZATION_NAME_ETAG, eTag)
+            put(SERIALIZATION_NAME_HTTPRESULT, httpResult.serialize())
+        }.toString()
+    }
+
+    companion object {
+        fun deserialize(serialized: String): HTTPResultWithETag {
+            val jsonObject = JSONObject(serialized)
+            val eTag = jsonObject.getString(SERIALIZATION_NAME_ETAG)
+            val serializedHTTPResult = jsonObject.getString(SERIALIZATION_NAME_HTTPRESULT)
+            return HTTPResultWithETag(eTag, HTTPResult.deserialize(serializedHTTPResult))
+        }
+    }
+}
+
+class ETagManager(
+    private val prefs: SharedPreferences
+) {
+    private var hashesByHttpRequest: Map<HTTPRequest, RequestHash> = emptyMap()
+
+    internal fun addETagHeaderToRequest(
+        httpRequest: HTTPRequest,
+        refreshETag: Boolean = false
+    ): HTTPRequest {
+        val eTagHeader = ETAG_HEADER_NAME to if (refreshETag) "" else getETag(httpRequest)
+        val updatedHeaders = httpRequest.headers + mapOf(eTagHeader)
+        return HTTPRequest(httpRequest.fullURL, updatedHeaders, httpRequest.body)
+    }
+
+    internal fun processResponse(
+        httpRequest: HTTPRequest,
+        connection: HttpURLConnection,
+        result: HTTPResult
+    ): HTTPResult? {
+        val eTagInResponse = connection.getHeaderField(ETAG_HEADER_NAME)
+        if (eTagInResponse != null) {
+            if (result.responseCode == NOT_MODIFIED_RESPONSE_CODE) {
+                return getStoredResult(httpRequest)?.httpResult
```

Comment:
maybe this is the best line to log of something potentially being wrong: if the response code is `NOT_MODIFIED_RESPONSE_CODE` but `getStoredResult(httpRequest)` returns `null`, we might wanna log that we were expecting to be able to use a cached response, but it can't be found now